uvicorn/supervisors/basereload.py

```diff
@@ -20,21 +20,21 @@ def __init__(self, config, target, sockets):
         self.config = config
         self.target = target
         self.sockets = sockets
+        self.processes = []
         self.should_exit = threading.Event()
         self.pid = os.getpid()
 
     def signal_handler(self, sig, frame):
         """
         A signal handler that is registered with the parent process.
         """
-        self.should_exit.set()
+        for process in self.processes:
+            try:
+                shutdown_subprocess(process.pid)
+            except Exception as exc:
+                logger.error(f"Could not stop child process {process.pid}: {exc}")
```

Comment:
The `self.processes` attribute is not necessary IMO, one alternative is to check for `self.process` being set:

```suggestion
        if hasattr(self, 'process'):
            try:
                shutdown_subprocess(self.process.pid)
            except Exception as exc:
                logger.error(f"Could not stop child process {self.process.pid}: {exc}")
```

or (cleaner), add `self.process = None` in the constructor, and check for `if self.process is not None: ...` instead.

WDYT?

Reply:
yep